### PR TITLE
Phase field fracture displacement vector

### LINKED ISSUE #8197: PF fracture model issue: c goes above 1
### Description of the enhancement or error report
In the phase field fracture model, we usually want to input displacement as a vector instead of two or three scalars, so we modified the corresponding codes to input the displacement as a vector input.
Also, when we run the simulation, the c goes above 1. 
### Rationale for the enhancement or information for reproducing the error
Usually c should be less than or equal to 1, and stress is a function of c and goes to minimum value when c=1, also the stress would increase when c is larger than 1. Thus, it is necessary for us to fix this issue to get the correct stress distribution. In order to fix this issue, we added one new parameter p=1-c in the kernel code file in which p=0 when c goes above 1, this could make sure that the stress would not increase when c keeps going above 1.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

### PR DESCRIPTION
**Pull request by zhangsf2015**
Closes #8197

### CODE REVIEW

File: modules/combined/tests/phase_field_fracture_viscoplastic/crack2d.i [hunk 1]
```diff
@@ -226,5 +205,6 @@
 []
 
 [Outputs]
+  file_base = crack_2d_out
```
Comment (dschwen) [inline modules/combined/tests/phase_field_fracture_viscoplastic/crack2d.i]: Please do not use file base.

File: modules/phase_field/include/kernels/PFFracBulkRate.h [hunk 2]
```diff
@@ -47,14 +47,12 @@ class PFFracBulkRate : public KernelValue
   ///Auxiliary variable: beta = Laplacian of c
   const VariableValue & _betaval;
   const unsigned int _beta_var;
+  // std::string _base_name;
```
Comment (dschwen) [inline modules/phase_field/include/kernels/PFFracBulkRate.h]: Remove, it is uncommented below

File: modules/phase_field/include/kernels/PFFracBulkRate.h [hunk 3]
```diff
@@ -47,14 +47,12 @@ class PFFracBulkRate : public KernelValue
...
 
-  const bool _xdisp_coupled;
-  const bool _ydisp_coupled;
-  const bool _zdisp_coupled;
-
-  const unsigned int _xdisp_var;
-  const unsigned int _ydisp_var;
-  const unsigned int _zdisp_var;
+  /// Coupled displacement variables
+  unsigned int _ndisp;
```
Comment (dschwen) [inline modules/phase_field/include/kernels/PFFracBulkRate.h]: Const

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 4]
```diff
@@ -31,16 +31,19 @@ PFFracBulkRate::PFFracBulkRate(const InputParameters & parameters):
...
   _beta_var(coupled("beta")),
-  _xdisp_coupled(isCoupled("disp_x")),
-  _ydisp_coupled(isCoupled("disp_y")),
-  _zdisp_coupled(isCoupled("disp_z")),
-  _xdisp_var(_xdisp_coupled ? coupled("disp_x") : 0),
-  _ydisp_var(_ydisp_coupled ? coupled("disp_y") : 0),
-  _zdisp_var(_zdisp_coupled ? coupled("disp_z") : 0),
+
+  _ndisp(coupledComponents("displacements")),
+  _disp_var(3),
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: Size with _ndisp

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 5]
```diff
@@ -31,16 +31,19 @@ PFFracBulkRate::PFFracBulkRate(const InputParameters & parameters):
...
+  _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+
   _l(getParam<Real>("l")),
   _visco(getParam<Real>("visco"))
-{
-}
+
+  {
+    for (unsigned int i = 0; i < _ndisp; ++i)
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: Indentation

**Comment by dschwen** [comment]
Please use a more meaningful pull request title and move the issue number in the pr body.

**Comment by dschwen** [comment]
Pleas squash your first five commits together and clean up the commit messages. Look at other commits in the MOOSE repository. No need to repeat _"Issue number"_ or mention the names of the files you change. Please describe _what_ the change does.

File: modules/phase_field/include/kernels/PFFracBulkRate.h (hunk 2 above)
Comment (zhangsf2015) [inline modules/phase_field/include/kernels/PFFracBulkRate.h]: Okay, I will remove it. Thanks.

File: modules/phase_field/include/kernels/PFFracBulkRate.h (hunk 3 above)
Comment (zhangsf2015) [inline modules/phase_field/include/kernels/PFFracBulkRate.h]: Okay. Thank you!

File: modules/phase_field/src/kernels/PFFracBulkRate.C (hunk 4 above)
Comment (zhangsf2015) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: That's better. Thanks!

File: modules/phase_field/src/kernels/PFFracBulkRate.C (hunk 5 above)
Comment (zhangsf2015) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: It helps. Thank you.

File: modules/combined/tests/phase_field_fracture_viscoplastic/crack2d.i (hunk 1 above)
Comment (zhangsf2015) [inline modules/combined/tests/phase_field_fracture_viscoplastic/crack2d.i]: Okay, I will fix it.

File: modules/combined/tests/phase_field_fracture/void2d.i [hunk 6]
```diff
@@ -42,21 +42,20 @@
...
     component = 0
     save_in = resid_x
     c = c
   [../]
   [./solid_y]
     type = StressDivergencePFFracTensors
-    variable = v
-    displacements = 'u v'
+    variable = disp_y
+    displacements = 'disp_x disp_y'
```
Comment (dschwen) [inline modules/combined/tests/phase_field_fracture/void2d.i]: I suggest moving those to GlobalParams as well

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 7]
```diff
@@ -53,14 +56,16 @@ PFFracBulkRate::computeDFDOP(PFFunctionType type)
     {
       //Order parameter for damage variable
       Real c = _u[_qp];
-      Real x =   _l * _betaval[_qp] + 2.0*(1.0-c) * _G0_pos[_qp]/gc - c/_l;
+      Real x =   _l * _betaval[_qp] + 2.0*(1-c) * _G0_pos[_qp]/gc - c/_l;
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: undo this change and add whitespace around the operators

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 8]
```diff
@@ -93,33 +98,27 @@ PFFracBulkRate::computeQpOffDiagJacobian(unsigned int jvar)
   Real c = _u[_qp];
   Real gc = _gc_prop[_qp];
 
-  Real x = _l * _betaval[_qp] + 2.0*(1.0-c) * (_G0_pos[_qp]/gc) - c/_l;
+
+  Real x = _l * _betaval[_qp] + 2.0*(1 - c) * (_G0_pos[_qp]/gc) - c/_l;
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: undo all changes `1.0` -> `1`. The one is not an integer and I would very much prefer if we explicitly add the decimal `.0` to floating point literals.

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 9]
```diff
@@ -93,33 +98,27 @@ PFFracBulkRate::computeQpOffDiagJacobian(unsigned int jvar)
...
+    if (jvar == _beta_var)
+      return xfacbeta * _phi[_j][_qp] * _test[_i][_qp];
+    else if (jvar == _disp_var[k])
+    {
+      c_comp = k;
+      disp_flag = true;
+    }
+    else
+      return 0.0;
+  // // Contribution of displacements to off diag Jacobian of c
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: remove one `//`

File: modules/tensor_mechanics/src/materials/LinearIsoElasticPFDamage.C [hunk 10]
```diff
@@ -84,7 +86,10 @@ LinearIsoElasticPFDamage::updateVar()
   //Used in PFFracBulkRate Jacobian
   _dG0_pos_dstrain[_qp] = stress0pos;
   //Used in StressDivergencePFFracTensors Jacobian
-  _dstress_dc[_qp] = -stress0pos * (2.0 * (1.0 - c));
+  if (c < 1.0)
+    _dstress_dc[_qp] = -stress0pos * (2.0 * (1.0 - c));
+  else
+    _dstress_dc[_qp] = -stress0pos * 0;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/LinearIsoElasticPFDamage.C]: just `0.0`

**Comment by dschwen** [comment]
:+1: for changing this over to the `displacements` vector coupling! I'll take another look once you address the comments.

File: modules/combined/tests/phase_field_fracture/void2d.i (hunk 6 above)
Comment (zhangsf2015) [inline modules/combined/tests/phase_field_fracture/void2d.i]: Okay, no problem

File: modules/combined/tests/phase_field_fracture/void2d.i [hunk 11]
```diff
@@ -178,11 +177,12 @@
   nl_max_its = 10
 
   dt = 1e-4
-  dtmin = 1e-4
+  dtmin = 1e-5
   num_steps = 2
 []
 
 [Outputs]
+  file_base = void2d_out
```
Comment (zhangsf2015) [inline modules/combined/tests/phase_field_fracture/void2d.i]: Sure, thanks.

File: modules/phase_field/src/kernels/PFFracBulkRate.C (hunk 7 above)
Comment (zhangsf2015) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: Great, it is true. Thanks

File: modules/phase_field/src/kernels/PFFracBulkRate.C (hunk 9 above)
Comment (zhangsf2015) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: Okay. No problem

File: modules/tensor_mechanics/src/materials/LinearIsoElasticPFDamage.C (hunk 10 above)
Comment (zhangsf2015) [inline modules/tensor_mechanics/src/materials/LinearIsoElasticPFDamage.C]: Okay, I will try this.

File: framework/include/utils/RankTwoTensor.h [hunk 12]
```diff
@@ -170,6 +170,9 @@ class RankTwoTensor
   RankTwoTensor operator - () const;
 
   /// performs _vals *= a
```
Comment (dschwen) [inline framework/include/utils/RankTwoTensor.h]: Please fix the comment to actually describe the operator. I'm not convinced that we should implement this overload at all. It is not obvious what it should do. Should it maybe rather set the tensor to I*a (I being the identity matrix)? 

Are you using this for anything other than `tensor = 0.0`? If not then please use `tensor.zero()` and remove this method.

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 13]
```diff
@@ -31,16 +31,19 @@ PFFracBulkRate::PFFracBulkRate(const InputParameters & parameters):
...
+
   _l(getParam<Real>("l")),
   _visco(getParam<Real>("visco"))
-{
-}
+
+  {
+      for (unsigned int i = 0; i < _ndisp; ++i)
+      _disp_var[i] = coupled("displacements", i);
+  }
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: indentation is off here as well

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 14]
```diff
@@ -55,12 +58,14 @@ PFFracBulkRate::computeDFDOP(PFFunctionType type)
       Real c = _u[_qp];
       Real x =   _l * _betaval[_qp] + 2.0*(1.0-c) * _G0_pos[_qp]/gc - c/_l;
 
+
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: rm extra blank line

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 15]
```diff
@@ -55,12 +58,14 @@ PFFracBulkRate::computeDFDOP(PFFunctionType type)
...
       return -(( std::abs(x) + x )/2.0)/_visco;
     }
     case Jacobian:
     {
       Real c = _u[_qp];
-      Real x = _l * _betaval[_qp] + 2.0 * (1.0 - c) * _G0_pos[_qp] / gc - c / _l;
+
+      Real x = _l * _betaval[_qp] + 2.0 * (1.0-c) * _G0_pos[_qp] / gc - c / _l;
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: operator spacing around `-`

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 16]
```diff
@@ -93,33 +98,27 @@ PFFracBulkRate::computeQpOffDiagJacobian(unsigned int jvar)
...
+    if (jvar == _beta_var)
+      return xfacbeta * _phi[_j][_qp] * _test[_i][_qp];
+    else if (jvar == _disp_var[k])
+    {
+      c_comp = k;
+      disp_flag = true;
+    }
+    else
+      return 0.0;
+  // Contribution of displacements to off diag Jacobian of c
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: add empty line above the comment

File: modules/phase_field/src/kernels/PFFracBulkRate.C [hunk 17]
```diff
@@ -93,33 +98,27 @@ PFFracBulkRate::computeQpOffDiagJacobian(unsigned int jvar)
...
-  }
-  else if (_zdisp_coupled && jvar == _zdisp_var)
-  {
-    c_comp = 2;
-    disp_flag = true;
-  }
-  else
-    return 0.0;
-  //Contribution of displacements to off diag Jacobian of c
+
```
Comment (dschwen) [inline modules/phase_field/src/kernels/PFFracBulkRate.C]: remove extra empty line here

File: modules/tensor_mechanics/src/materials/LinearIsoElasticPFDamage.C [hunk 18]
```diff
@@ -84,7 +86,10 @@ LinearIsoElasticPFDamage::updateVar()
   //Used in PFFracBulkRate Jacobian
   _dG0_pos_dstrain[_qp] = stress0pos;
   //Used in StressDivergencePFFracTensors Jacobian
-  _dstress_dc[_qp] = -stress0pos * (2.0 * (1.0 - c));
+  if (c < 1.0)
+    _dstress_dc[_qp] = -stress0pos * (2.0 * (1.0 - c));
+  else
+    _dstress_dc[_qp] = 0.0;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/LinearIsoElasticPFDamage.C]: use `_dstress_dc[_qp].zero()` here

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/include/utils/RankTwoTensor.h (hunk 12 above)
Comment (dschwen) [inline framework/include/utils/RankTwoTensor.h]: Actually I'm pretty convinced that we should _not_ add this operator overload @tonkmr

**Comment by dschwen** [comment]
Closing in favor of #8227, which has the requested fixes added.

**Pull request closed without merging**